### PR TITLE
Allow deployment of data projects to public networks

### PR DESCRIPTION
**Pull request by russellseymour**
#### 📲 What

Updated the code and the pipelines so that it is possible to deploy the Ensono Stacks data project to private and public networks.

Fixed an issue where it was not possible to tear down and rebuild an enviornment using the pipeline and Terraform.

When constantly redeploying resources it became apparent that there were issues with Key Vaults and SQL Server instances. Microsoft have made a change that means Key Vaults can only be soft deleted, the only setting that can be configured is the minimum amount of time it can stay softw deleted, which can be no shorter than 7 days. SQL Server instances are deleted behind the scenes, even though you do not see them in the portal.

#### 🤔 Why

When the project was tested in December 23 it became evident that it was not possible to deploy the infrastrastructure and Databricks pipelines to a public network - it only worked if a set of Virtual Networks and subnets had been deployed to allow for private communcation between the components.

It is necessary to be able to tear down an environment and rebuild it occasionally to check that all of the components still work together after incremental updates.

With Key Vault and SQL Server, as the resources are still lingering subsequent immediate deployments can fail. This is down to the same name for the resource being set, in which case Terraform sees that a KV of the same name is soft-deleted it will bring it back. When it comes to the SQL Server, the deployment will time out after 30 minutes as it is attempting to deploy an instance with the same name as one that is currently being deleted.

#### 🛠 How

The infrastructure code did not have the necessary flags and guards to deploy resources without a private network. For example it expected the names of existing virtual networks and subnets. This has been modified so that if the `enable_private_networks` is set to false, the components that need to connect to a private network are not executed,

The Databricks provider for Terraform has a bug which does not allow Terraform to tear down the Databricks resource. Therefore is not possible to remove the resource group that all the resources have been deployed into. An issue has been raised on the GitHub repo - https://github.com/databricks/terraform-provider-databricks/issues/3137 - but it has not been looked into.

To get around this problem I have written a PowerShell script that interrogates the deployment infrastructure and determines what needs to be removed from shared components, e.g. subnets. It will then remove the state blob in storage and then remove the resources from Azure by deleting the resource group.

By adding a random seed to the KV and SQL Instances it is possible to change the name of the resource so that when it is torn down and redeployed there are no conflicts. The random seed is kept for the entire duration of the life of the resources and is kept in state so the resources are not destroyed / rebuilt on subsequent runs.

#### 👀 Evidence

Screenshot of resources in Azure running without a private network
image

#### 🕵️ How to test

Notes for QA

#### ✅ Acceptance criteria Checklist

- [ ] Code peer reviewed?
- [ ] Documentation has been updated to reflect the changes?
- [ ] Passing all automated tests, including a successful deployment?
- [ ] Passing any exploratory testing?
- [ ] Rebased/merged with latest changes from development and re-tested?
- [ ] Meeting the Coding Standards?

### CODE REVIEW

File: deploy/azure/infra/main.tf [hunk 1]
```diff
@@ -35,28 +44,49 @@ resource "azurerm_resource_group" "default" {
...
+    EOT
+  }
+
+  depends_on = [module.kv_default]
+}
+*/
+
 # module call for ADF
 module "adf" {
   source                          = "git::https://github.com/ensono/stacks-terraform//azurerm/modules/azurerm-adf?ref=master"
```
Comment (ElvenSpellmaker) [inline deploy/azure/infra/main.tf:92]: Pin to version

File: deploy/azure/infra/main.tf [hunk 2]
```diff
@@ -120,14 +156,19 @@ resource "azurerm_data_factory_managed_private_endpoint" "db_auth_pe" {
...
+  /*
+  {
     blob = module.adls_default.storage_account_ids[0]
     adls = module.adls_default.storage_account_ids[1]
     kv   = module.kv_default.id
     sql  = module.sql.sql_server_id
     adb  = module.adb.adb_databricks_id
     # Add more resources as needed
   }
+  */
```
[the remark this answers is not in the document]
Comment (ElvenSpellmaker) [reply deploy/azure/infra/main.tf:170]: Remove

File: deploy/azure/infra/main.tf [hunk 3]
```diff
@@ -239,43 +280,42 @@ module "adls_default" {
...
-  dns_resource_group_name       = var.dns_resource_group_name
+  dns_resource_group_name       = local.dns_zone_resource_group_name
   public_network_access_enabled = var.sql_public_network_access_enabled
   //As the default SKU in the module is basic, we need to set this to 0 otherwise it defaults to 60 and never gets applied.
   auto_pause_delay_in_minutes = 0
 }
 
 module "adb" {
-  source                                   = "git::https://github.com/ensono/stacks-terraform//azurerm/modules/azurerm-adb?ref=master"
+  source                                   = "git::https://github.com/ensono/stacks-terraform//azurerm/modules/azurerm-adb?ref=feature/kv-purge"
```
Comment (ElvenSpellmaker) [inline deploy/azure/infra/main.tf:318]: Pin to tag

File: deploy/azure/infra/provider.tf [hunk 4]
```diff
@@ -8,8 +8,8 @@ provider "azurerm" {
 }
 
 provider "databricks" {
-  host                        = module.adb.databricks_hosturl
-  azure_workspace_resource_id = module.adb.adb_databricks_id
+  host                        = module.adb.databricks_hosturl != "" ? module.adb.databricks_hosturl : var.adb_databricks_hosturl
+  azure_workspace_resource_id = module.adb.adb_databricks_id != "" ? module.adb.adb_databricks_id : var.adb_databricks_id
```
Comment (ElvenSpellmaker) [inline deploy/azure/infra/provider.tf:12]: Only use module?

File: deploy/azure/infra/vars.tf [hunk 5]
```diff
@@ -540,3 +522,13 @@ variable "sql_public_network_access_enabled" {
...
+
+variable "adb_databricks_hosturl" {
+  type    = string
+  default = ""
+}
+
+variable "adb_databricks_id" {
+  type    = string
+  default = ""
+}
```
Comment (ElvenSpellmaker) [inline deploy/azure/infra/vars.tf:534]: Remove these

File: deploy/azure/networking/locals.tf [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    uksouth       = "uks"
+    ukwest        = "ukw"
+    eastus        = "use"
+    eastus2       = "use2"
+    westus        = "usw"
+    eastasia      = "ase"
+    southeastasia = "asse"
+  }
+
+  network_details = {
```
Comment (ElvenSpellmaker) [inline deploy/azure/networking/locals.tf:16]: Add details to the readme please

File: build/azDevOps/azure/network/air-infrastructure-data-network.yml [hunk 7]
```diff
@@ -13,11 +42,12 @@ variables:
   - name: tf_state_key
     value: $(network_domain)
   - name: TF_FILE_LOCATION
-    value: $(self_repo_tf_src)
+    value: /app/deploy/azure/networking
```
[the remark this answers is not in the document]
Comment (russellseymour) [reply build/azDevOps/azure/network/air-infrastructure-data-network.yml:45]: Corrected

File: build/common-vars.yml [hunk 8]
```diff
@@ -2,12 +2,12 @@ variables:
   - name: region
     value: westeurope
   - name: company
-    value: amido
+    value: beaker
```
[the remark this answers is not in the document]
Comment (russellseymour) [reply build/common-vars.yml:5]: Modified to a new fake name and project

File: build/taskctl/tasks.yaml [hunk 9]
```diff
@@ -34,20 +34,42 @@ tasks:
...
+    description: Create Terraform variables file
+    command:
+      - /app/build/scripts/Set-TFVars.ps1 | Tee-Object -FilePath "${env:TF_FILE_LOCATION}/terraform.tfvars"
+
   infra:plan:
     context: powershell
     description: Plan Terraform
     command:
-      (Invoke-Terraform -Plan -Path $env:TF_FILE_LOCATION -Arguments "-input=false","-out=tfplan" -Debug) -replace "{{"{{"}}", "{{"{{"}}`"{{"{{"}}`"}}"
+      # - (Invoke-Terraform -Plan -Path $env:TF_FILE_LOCATION -Arguments "-input=false","-out=tfplan" -Debug) -replace "{{"{{"}}", "{{"{{"}}`"{{"{{"}}`"}}"
```
[the remark this answers is not in the document]
Comment (russellseymour) [reply build/taskctl/tasks.yaml:50]: Removed commented commands

File: build/azDevOps/azure/network/air-infrastructure-data-network.yml [hunk 10]
```diff
@@ -44,65 +74,86 @@ stages:
...
+                        TaskctlVersion: ${{ variables.TaskctlVersion }}
+
+                    # Create a tfvars file for the values
+                    # This is so that variables can be set and reused as required
+                    - task: Bash@3
+                      displayName: Create Terraform variables file
+                      inputs:
+                        targetType: inline
+                        script: |
+                          taskctl infra:vars
```
[the remark this answers is not in the document]
Comment (russellseymour) [reply build/azDevOps/azure/network/air-infrastructure-data-network.yml:109]: Yes this creates the tfvars file that the variables are read from

File: build/scripts/Remove-Environment.ps1 [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    NetworkResourceGroupName = $env:vnet_resource_group_name
+    VirtualNetworkName = $env:vnet_name
+    TerraformStorageAccount = $env:tf_state_storage
+    TerraformStateKey = $env:tf_state_key
+    TerraformWorkspace = $env:ENV_NAME
+    ResourceGroupName = $rg_name
+    subnets = $env:subnets
+}
+
+Remove-Environment @splat
```
[the remark this answers is not in the document]
Comment (russellseymour) [reply build/scripts/Remove-Environment.ps1:145]: This is required to get around the issue of databricks preventing the infrastrtcure bneing torn down by Terraform.

File: build/azDevOps/azure/network/air-infrastructure-data-network.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (ElvenSpellmaker) [reply build/azDevOps/azure/network/air-infrastructure-data-network.yml:109]: Yeah we had the discussion over making this eventually read a `variables.tf` and only creating a tfvars based on that. That's what I mean by clever.

**Review by ElvenSpellmaker — APPROVED** [review]

**Pull request merged**